tinygrad/ops.py

```diff
@@ -18,19 +20,41 @@ class LoadOps(Enum): EMPTY = auto(); RAND = auto(); CONST = auto(); FROM = auto(
 Op = Union[UnaryOps, BinaryOps, ReduceOps, MovementOps, LoadOps, FusedOps]
 OpType = Union[Type[UnaryOps], Type[BinaryOps], Type[ReduceOps], Type[MovementOps], Type[LoadOps], Type[FusedOps]]
 
-class LazyOp(NamedTuple):
-  op: Op
-  # Any == Union[LazyOp, LazyBuffer, DeviceBuffer]
-  src: Tuple[Any, ...]  # type: ignore
-  arg: Any = None
+class LazyOp:
   # TODO: add dest to support multiple outputs. on second thought, multiple outputs will have multiple LazyOps.
-
-# Any == Union[LazyBuffer, DeviceBuffer]
-def get_buffers(op:LazyOp) -> List[Any]: return functools.reduce(operator.add, [get_buffers(x) if isinstance(x, LazyOp) else [x] for x in op.src], [])
-def get_lazyops(op:LazyOp) -> List[LazyOp]: return functools.reduce(operator.add, [get_lazyops(x) for x in op.src if isinstance(x, LazyOp)], [op])
-def map_buffers(real_srcs:Dict[Any, Any], x:Any) -> LazyOp:
-  if len(real_srcs) and x in real_srcs: return map_buffers(real_srcs, real_srcs[x]) if isinstance(real_srcs[x], LazyOp) else real_srcs[x]
-  return LazyOp(x.op, tuple((map_buffers(real_srcs, y) if isinstance(y, LazyOp) else real_srcs[y]) for y in x.src), x.arg)
+  __slots__ = "op", "src", "arg", "buffers", "__weakref__"
+  op: Op
+  src: Tuple[Union[LazyOp, LazyBuffer], ...]
+  arg: Union[LazyOp, LazyBuffer]
+  buffers: Tuple[LazyBuffer]
```

Comment:
I don't like that this is storing state in two places. What if this doesn't match src?

Reply:
That can't happen by construction

We never delete buffers from the AST and ops are always replaced, never modified in-place so the final computational graph will always be consistent

Reply:
Sure, LazyOps are immutable. But my concern is an invalid LazyOp will be constructed where the state doesn't match. There's a few places where they are created, and it's so easy to write a bug. Can't the class just cache the buffers on first access? Or even on construction.

Reply:
On construction would work, we can keep the same `__init__` and just remove the ability to pass in buffers in the constructor, would that alleviate your concerns?

The option to pass in the buffers to the constructor was added because for pretty much all internal operations in `lazy.py`, we know in advance what the buffers are since we're just shuffling around nodes in the AST and so we don't need the additional recursion in the constructor

Reply:
Yea, that works. Then it can't have a bug somewhere with a mismatch.